File: Client/Assets/HistoryStateHelper.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

(function() {
    var nativeHistoryPushState = window.history.pushState;
    var nativeHistoryReplaceState = window.history.replaceState;

    // We need to catch calls to `history.pushState()` in order to
    // notify the BrowserViewController so that history can be
    // recorded in single-page web applications that do all rendering
    // on the client side.
    window.history.pushState = function(state, title, url) {
        nativeHistoryPushState.apply(this, arguments);
        webkit.messageHandlers.historyStateHelper.postMessage({
            pushState: true,
            state: state,
            title: title,
            url: url
        });
    };

    // We need to catch calls to `history.replaceState()` in order to
    // notify the BrowserViewController so that history can be
    // recorded in single-page web applications that do all rendering
    // on the client side.
    window.history.replaceState = function(state, title, url) {
      nativeHistoryReplaceState.apply(this, arguments);
      webkit.messageHandlers.historyStateHelper.postMessage({
          replaceState: true,
          state: state,
          title: title,
          url: url
      });
    };
})();

File: Client/Frontend/Browser/HistoryStateHelper.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Foundation
import WebKit

protocol HistoryStateHelperDelegate: class {
    func historyStateHelper(_ historyStateHelper: HistoryStateHelper, didPushOrReplaceStateInTab tab: Tab)
}

// This tab helper is needed for injecting a user script into the
// WKWebView that intercepts calls to `history.pushState()` and
// `history.replaceState()` so that the BrowserViewController is
// notified when the user navigates a single-page web application.
class HistoryStateHelper: TabContentScript {
    weak var delegate: HistoryStateHelperDelegate?
    fileprivate weak var tab: Tab?
    
    required init(tab: Tab) {
        self.tab = tab
    }
    
    func scriptMessageHandlerName() -> String? {
        return "historyStateHelper"
    }
    
    func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
        if let tab = tab {
            DispatchQueue.main.async {
                self.delegate?.historyStateHelper(self, didPushOrReplaceStateInTab: tab)
            }
        }
    }
    
    class func name() -> String {
        return "HistoryStateHelper"
    }
}
